src/sctools/bam.py

```diff
@@ -115,11 +119,95 @@ def tag(self, output_bam_name, tag_generators):
           (see fastq.Tag)
         """
         with pysam.AlignmentFile(self.bam_file, 'rb', check_sq=False) as inbam, \
-                pysam.AlignmentFile(output_bam_name, 'wb', header=inbam.header) as outbam:
+                pysam.AlignmentFile(output_bam_name, 'wb', template=inbam) as outbam:
 
             # zip up all the iterators
             for *tag_sets, sam_record in zip(*tag_generators, inbam):
                 for tag_set in tag_sets:
                     for tag in tag_set:
                         sam_record.set_tag(*tag)
```

Comment:
Don't know what size of the exact data here. These nested loops look a little bit expensive, but if the data is relatively small, this would be fine.